automation/tests-storage.sh

```diff
@@ -12,28 +33,107 @@ create_loop_devices() {
 }
 
 setup_storage() {
-    python3 tests/storage/userstorage.py setup
+    # Configure lvm to ignore udev events, otherwise some lvm tests hang.
+    mkdir -p /etc/lvm
+    cp docker/lvmlocal.conf /etc/lvm/
+
+    # Make sure we have enough loop device nodes. Using 16 devices since with 8
+    # devices we have random mount failures.
+    create_loop_devices 16
+
+    # Build vdsm.
+    ./autogen.sh --system
+    make
+
+    # Setup user stoage during the tests.
+    make storage
 }
 
 teardown_storage() {
-    python3 tests/storage/userstorage.py teardown \
+    make clean-storage \
         || echo "WARNING: Ingoring error while tearing down user storage"
 }
 
-# Configure lvm to ignore udev events, otherwise some lvm tests hang.
-mkdir -p /etc/lvm
-cp docker/lvmlocal.conf /etc/lvm/
+# Process command-line arguments
+while test $# -gt 0
+    do
+        case $1 in
 
-# Make sure we have enough loop device nodes. Using 16 devices since with 8
-# devices we have random mount failures.
-create_loop_devices 16
+            root )
 
-# Build vdsm.
-./autogen.sh --system
-make
+                mode=root
+                ;;
+            
+            vdsm )
+
+                mode=vdsm
+                ;;
+
+            -u=* | --user=* )
+
+                user="${1#*=}"
+                if [ ! -z ${mode+x} ]; then
+                    mode=vdsm
+                fi
+                ;;
+            
+            -u | --user )
+
+                shift
+                if [ -z $1 ]; then
+                    echo "error: user expected after --user"
+                    exit 1
+                fi
+                user="$1"
+                if [ ! -z ${mode+x} ]; then
+                    mode=vdsm
+                fi
+                ;;
+            
+            * )
+
+                echo "Unknown option: $1"
+                print_help
+                exit 1
+            ;;
+
+        esac
+        shift
+done
+
+# Defaults if not provided
+if [ ! -z ${user+x} ]; then
+    user=root
+fi
+if [ ! -z ${mode+x} ]; then
+    mode=root
+fi
+
+if [ "$user" != "root" -a "$mode" = "root" ]; then
+    echo "error: non-privileged user ($user) can not run tests marked as root"
+    print_help
+    exit 1
+fi
+
+echo "Running $mode tests as user $user"
 
-# Setup user stoage during the tests.
-trap teardown_storage EXIT
 setup_storage
+if [ "$user" != "root" ]; then
+    # Change ownership of current and storage folders
+    # to allow non-privileged access.
+    chown -R vdsm:kvm ./ /var/tmp/vdsm*
+fi
+# Teardown storage before exit.
+trap teardown_storage EXIT
 
-make tests-storage
+if [ "$mode" = "root" ]; then
+    # root mode
+    # Run only tests marked as root
+    tox -e storage -- --cov-fail-under 47 \
```

Comment:
The coverage value is better managed as configuration value instead of hiding
in the body of the script.

The implementation should be as a separate tox env instead of duplicating
most of the testenv:storage here.

Implementing the logic here means there is no good way to run the storage 
tests directly with tox - you must use this wrapper script. If we have a tox
env that does everything, a developer can run the tests as normal user
using:

    $ tox -e storage-user

And the root tests as:

    # tox -e storage-root

So the only thing the script has to do is to change the user
before running tox and run the right target.

Reply:
Ok, it makes sense to me.

I just wasn't sure if this is something we wanted in tox.ini as these environments are practically only going to be used in the pipelines. But I'll add them.

Reply:
Why tox env are not for local use? Today we use `tox -e storage` and it skip the tests that needs root and fail when coverage is too low because the root tests were skipped.

With this change we can have:
- tox -e storage or tox -e storage-user
- tox -e storage-root

Both will run only the relevant tests and give correct coverage results. Looks like great improvement even if this was not used in the CI.

One issue with separate targets is how to get complete coverage of all tests - both user and root tests, but this can solved by another make target combining both tox envs and generating combined coverage.

The root target is basically workaround since almost no code in storage needs root, we actually need supervdsm running to run code as root, but we don't have this test infa now. When we will have it, we could remove the storage-root target and have a single target.

Reply:
I didn’t mean it as it is not good, but I though that in general we would keep using the storage env for local testing. But yeah, on a second thought, this will useful for both the CIs and local.

The make target to combine coverage reports I will leave for the next patch.